specs/dwritecore/DWriteCore.md

```diff
@@ -0,0 +1,255 @@
+interface that an application can use to access the bitmap data without going through GDI.
+
+# Examples
+
+## Creating a restricted factory
+
+An application calls the [DWriteCreateFactory](https://docs.microsoft.com/en-us/windows/win32/api/dwrite/nf-dwrite-dwritecreatefactory)
+function to create a factory object, which is the entry-point to all other DirectWrite APIs. An application can
+create a _restricted factory_ if it does not want the factory or objects created from it to read from or write
+to any cross-process or persistent cache. In addition, a restricted factory's system font collection only includes
+well-known system fonts.
+
+The following example creates a restricted factory.
+
+```c++
+ComPtr<IDWriteFactory7> spFactory;
+HRESULT hr = DWriteCreateFactory(
+    DWRITE_FACTORY_TYPE_RESTRICTED,
+    __uuidof(IDWriteFactory7),
+    (IUnknown**)&spFactory
+    );
+```
+
+## Drawing glyphs to a system memory bitmap
+
+An application creates a bitmap render target by calling [IDWriteGdiInterop::CreateBitmapRenderTarget](https://docs.microsoft.com/en-us/windows/win32/api/dwrite/nf-dwrite-idwritegdiinterop-createbitmaprendertarget).
+A bitmap render target encapsulates a bitmap in system memory, and eanbles rendering glyphs to the bitmap. On Windows,
+the bitmap is actually a GDI DIB selected into a memory HDC, and it is possible to get the bitmap pixels using GDI
+functions. However, DWriteCore introduces a simpler method through the new IDWriteBitmapRenderTarget2 interface.
+The following example shows both the old way and the new way.
+
+```c++
+DWRITE_BITMAP_DATA_BGRA32 TextRenderer::GetBitmapData(IDWriteBitmapRenderTarget* renderTarget)
+{
+    DWRITE_BITMAP_DATA_BGRA32 result = {};
+
+    // Query the bitmap render target for the new interface.
+    ComPtr<IDWriteBitmapRenderTarget2> renderTarget2;
+    HRESULT hr = renderTarget->QueryInterface(&renderTarget2);
+    if (SUCCEEDED(hr))
+    {
+        // IDWriteBitmapRenderTarget2 exists, so we can get the bitmap the easy way.
+        hr = renderTarget2->GetBitmapData(OUT &result);
+        if (FAILED(hr))
+        {
+            throw Exception{ hr };
+        }
+    }
+    else
+    {
+        // We're using an older version that doesn't implement IDWriteBitmapRenderTarget2,
+        // so we have to get the bitmap by going through GDI. First get the bitmap handle.
+        HDC hdc = renderTarget_->GetMemoryDC();
+        HGDIOBJ dibHandle = GetCurrentObject(hdc, OBJ_BITMAP);
+        if (dibHandle == nullptr)
+        {
+            throw Exception{ HRESULT_FROM_WIN32(GetLastError()) };
+        }
+
+        // Call a GDI function to fill in the DIBSECTION structure for the bitmap.
+        DIBSECTION dib;
+        if (!GetObjectW(dibHandle, sizeof(dib), &dib))
+        {
+            throw Exception{ HRESULT_FROM_WIN32(GetLastError()) };
+        }
+
+        result.width = dib.dsBm.bmWidth;
+        result.height = dib.dsBm.bmHeight;
```

Comment:
Top-down DIBs report their height as negative. We need to flip the sign before returning it. Otherwise, a 10-pixel tall bitmap is reported as 4 billion+ pixels tall.

```suggestion
        result.height = 0U-dib.dsBm.bmHeight;
```